daemon/disk_usage.go

```diff
@@ -5,46 +5,77 @@ import (
 	"fmt"
 	"sync/atomic"
 
+	"github.com/docker/docker/api/server/router/system"
 	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/errdefs"
 )
 
 // SystemDiskUsage returns information about the daemon data disk usage
-func (daemon *Daemon) SystemDiskUsage(ctx context.Context) (*types.DiskUsage, error) {
+func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsageOptions) (*types.DiskUsage, error) {
 	if !atomic.CompareAndSwapInt32(&daemon.diskUsageRunning, 0, 1) {
 		return nil, fmt.Errorf("a disk usage operation is already running")
 	}
 	defer atomic.StoreInt32(&daemon.diskUsageRunning, 0)
 
-	// Retrieve container list
-	allContainers, err := daemon.Containers(&types.ContainerListOptions{
-		Size: true,
-		All:  true,
-	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to retrieve container list: %v", err)
-	}
+	var (
+		containers []*types.Container
+		images     []*types.ImageSummary
+		layersSize int64
+		volumes    []*types.Volume
 
-	// Get all top images with extra attributes
-	allImages, err := daemon.imageService.Images(filters.NewArgs(), false, true)
-	if err != nil {
-		return nil, fmt.Errorf("failed to retrieve image list: %v", err)
+		err error
+	)
+	objectTypes := opts.ObjectTypes
+	if len(objectTypes) == 0 {
+		objectTypes = map[system.DiskUsageObject]struct{}{
```

Comment:
Nit/idea: I guess we no longer strictly need the intermediate variable, and could simplify to:

```suggestion
	if len(opts.ObjectTypes) == 0 {
		opts.ObjectTypes = map[system.DiskUsageObject]struct{}{
```